src/Microsoft.AspNetCore.JsonPatch/Helpers/DictionaryWrapper.cs

```diff
@@ -0,0 +1,100 @@
+        private readonly IDictionary<TKey, TValue> _targetDictionary;
+        private readonly bool useDynamic;
+
+        public DictionaryWrapper(IDictionary<TKey, TValue> targetDictionary)
+        {
+            _targetDictionary = targetDictionary;
+            KeyType = typeof(TKey);
+            ValueType = typeof(TValue);
+
+            // For the case when target dictionary is an ExpandoObject (which implements IDictionary<string, object>)
+            useDynamic = (KeyType == typeof(string)) && (ValueType == typeof(object));
+        }
+
+        public Type KeyType { get; }
+
+        public Type ValueType { get; }
+
+        public object GetValueForCaseInsensitiveKey(object key)
+        {
+            var foundKey = GetKeyUsingCaseInsensitiveSearch(key);
+            return _targetDictionary[foundKey];
+        }
+
+        public void SetValueForCaseInsensitiveKey(object key, object value)
+        {
+            var foundKey = GetKeyUsingCaseInsensitiveSearch(key);
+
+            // If the target object is an ExpandoObject, then the type of the value in the key-value pair is 'object'
+            // and so we cannot deserialize the raw json to a particular type.
+            // But if a value for a key already exists, then we try to figure out the type and deserialize the value to
+            // this type.
+            if (useDynamic && _targetDictionary.ContainsKey(foundKey) && _targetDictionary[foundKey] != null)
+            {
+                var result = ConversionHelper.ConvertToActualType(_targetDictionary[foundKey].GetType(), value);
+                if (result.CanBeConverted)
+                {
+                    _targetDictionary[foundKey] = CastTo<TValue>(result.ConvertedInstance);
+                    return;
+                }
+            }
+
+            _targetDictionary[foundKey] = CastTo<TValue>(value);
+        }
+
+        public void RemoveValueForCaseInsensitiveKey(object key)
+        {
+            var foundKey = GetKeyUsingCaseInsensitiveSearch(key);
+            _targetDictionary.Remove(foundKey);
+        }
+
+        public bool ContainsCaseInsensitiveKey(object key)
+        {
+            var foundKey = GetKeyUsingCaseInsensitiveSearch(key);
+            return _targetDictionary.ContainsKey(foundKey);
+        }
+
+        private TKey GetKeyUsingCaseInsensitiveSearch(object key)
+        {
+            // Example: a Guid key
+            if (KeyType != typeof(string))
+            {
+                var conversionResult = ConversionHelper.ConvertToActualType(KeyType, key);
+                if (!conversionResult.CanBeConverted)
+                {
+                    throw new JsonPatchException(
+                        message: Resources.FormatInvalidDictionaryKeyFormat(key, KeyType.FullName),
+                        innerException: null);
+                }
+                return CastTo<TKey>(conversionResult.ConvertedInstance);
+            }
+
+            var keyToFind = (string)key;
```

Comment:
> Can't key be a non-string but have a type conversion to string?

Not sure what you mean here, but why would a key be converted to a string by a dictionary? isn't hashcode used to store it.